### ISSUE TITLE
Is there a way to correlate that a detach or an end frame was received because the client initiated that?

### ISSUE THREAD

**Issue by amarzavery** [amqp/rhea#112]
I am trying to distinguish between the `*_close` event being raised due to the application/client closing it and the event being raised when the server sends the detach or an end frame (with or without an error). This is needed to accurately determine whether the entity needs to be recovered or not. At times the service sends an empty detach/end/close frame so I can't really rely on whether an error happened.

- Can this info be provided in the EventContext when the event is raised?
- Is it possible to cache the last frame that was sent by the sender/receiver/session in those objects (may be that can be a good indicator, not entirely sure)?

tldr: Some way to distinguish that the app/client initiated the close of link, session, connection would be really helpful.

**Comment by grs** [comment]
See https://github.com/amqp/rhea/pull/113 which I think will resolve this. WIth that fix you can use session.is_closed()/link.is_closed() to determine whether the close from the peer is a response to a locally initiated close request (the return value will be true if locally initiated, false otherwise).

**Comment by amarzavery** [comment]
That is awesome. Thanks a ton for fixing issues and implementing feature requests at jet speed 👍

**Comment by amarzavery** [comment]
fixed by #113
